### ISSUE TITLE
Can I use data parallel with different data on each machine?

### ISSUE THREAD

**Issue by jacopocecch** [microsoft/LightGBM#6148]
Hi, I was just wondering, can I use the LightGBM data parallel feature to simulate federated learning?
Suppose I have two machines and each machine has its local dataset, does it work? Or does the data in each machine have to be the same for it to work?
Thanks in advance!

**Comment by jameslamb** [comment]
> *Or does the data in each machine have to be the same for it to work?*

With data parallel training in LightGBM, each machine can hold a different non-overlapping subset of the full training data. This is described in some detail in these resources:

* https://lightgbm.readthedocs.io/en/stable/Features.html#optimization-in-distributed-learning
* one of the original LightGBM papers (["A Communication-Efficient Parallel Algorithm for Decision Tree"](https://papers.nips.cc/paper_files/paper/2016/hash/10a5ab2db37feedfdeaab192ead4ac0e-Abstract.html)).

> *can I use the LightGBM data parallel feature to simulate federated learning?*

It depends on how you define "federated learning".

I've usually seen that phrase used to imply that details about the dataset aren't shared between individual training processes, and that the only thing that's shared across the network are things like the learned parameters (or splits, for the case of tree-based stuff like LightGBM) and gradients.

I've also usually seen that phrase used to refer to doing distributing training in environments where there are restrictions on data movement ...e.g. you have training data in different geographic locations and do not want to physically move it, but want to train a model over all of it.

So I want to be sure you understand... LightGBM distributed training (all of them... data-parallel, feature-parallel, and voting-parallel) does involve a global sync-up where all training processes hold information like the **global** distribution of each feature (i.e., bin boundaries of the histograms). And that "distribution" includes specific values of categorical variables (which is a common representation for potentially sensitive information like country, zip code, age, etc.).

**Comment by jacopocecch** [comment]
@jameslamb Thank you for the answer, it's all clear now.

**Comment by shiyu1994** [comment]
@jacopocecch Thanks for using LightGBM.

> Suppose I have two machines and each machine has its local dataset, does it work? Or does the data in each machine have to be the same for it to work?

Yes, you can have different machine with its own part of data.

This parameter `pre_partition`
https://lightgbm.readthedocs.io/en/latest/Parameters.html#pre_partition
tells LightGBM whether the data in distributed training has been partitioned across machines or each machine has a full copy of the data.
In your case, you need to set `pre_partition=true`.

**Comment by jameslamb** [comment]
Ah thanks, I forgot about also setting `pre_partition=True`.

**Comment by jacopocecch** [comment]
@shiyu1994 thank you
